### ISSUE TITLE
Coalesce is too verbose? 

### ISSUE THREAD

**Issue by pik** [fish-shell/fish-shell#2504]
According to this document there are [no options available](http://hyperpolyglot.org/unix-shells) and according to this stack-exchange discussion [it doesn't look to nice either](http://unix.stackexchange.com/questions/81421/how-do-i-reference-a-variable-in-fish-shell-with-a-default-fallback). The best I can think of something like:

```
set -q $MY_PROGRAM_ROOT; and set program_root ~/.local/share/my_prorgram; \
or set program_root $MY_PROGRAM_ROOT
```

Is there a better alternative available?

**Comment by faho** [comment]
Nope, that's pretty much it. Like [Patrick Redmond says](http://unix.stackexchange.com/a/203847/55665) on that stack exchange thread, you can make a function that does it, though I'd make one small adjustment: 

Add the "--no-scope-shadowing" option to the definition, like this:

```fish
function orDefault --no-scope-shadowing
 set -q $argv[1]; and echo $$argv[1]; or echo $argv[2]
end
```

This will cause the function to inherit the current scope, which means it gets access to local variables as well.

The reason for this is in [fish's design document](http://fishshell.com/docs/current/design.html) as the "Law Of Orthogonality".

This does mean that fishscript is not always the shortest language, but is among the clearest (in my opinion, anyway). (This is similar to python's "explicit is better than implicit")

**Comment by zanchey** [comment]
I wonder about a flag to `set` to shorten this fairly common pattern.

**Comment by faho** [comment]
How'd the syntax look like on that?

`set var --default var1 default` or `set var --default $var1 default`? The latter'd need to ignore the expansion but would be a common mistake if we went with the former.

This is similar to an option to append or to append only if the value isn't included, mentioned in #2383.

I personally think shipping some functions might solve this.

**Comment by zanchey** [comment]
Perhaps `set --if-unset` and `set --if-empty`?

**Comment by faho** [comment]
But how do you pass the variable and the "else-value" to it?

Or do you mean "set a variable to this value only if it is unset/empty" i.e. avoid passing a second variable? That would solve the problem here a bit differently - you'd do

```fish
set --if-unset MY_PROGRAM_ROOT ~/.local/share/my_program
# and now you'd use MY_PROGRAM_ROOT directly
```

**Comment by zanchey** [comment]
Exactly the latter. The difference between `--if-unset` and `--if-empty` might trip people up, but then it already does; `set -q` is different to `test -n`.

**Comment by pik** [comment]
> The reason for this is in fish's design document as the "Law Of Orthogonality".

Generally I agree with this but here something that used to be very simple `${BAR:-default}`  has become extremely unruly (when multiple values need to be coalesced and then used in an expression). 

> Perhaps set --if-unset and set --if-empty?

Would be great to have shortflag version e.g. `set -c` for coalesce?

**Comment by zanchey** [comment]
There would certainly be short options. I don't know that "coalesce" is the right terminology - its use in this way is limited to SQL as far as I know.

**Comment by krader1961** [comment]
We should not introduce "syntactic sugar" for this pattern since it is not used often enough to justify the complexity; not to mention violating the "Law Of Orthogonality" as @faho points out. I would support adding a "--if-unset" option to the "set" command but not a "--if-empty" option. The latter case is even more rare and in my opinion should require being verbose (e.g., `if test -n $var`). I've seen too many instances where people have misused (probably inadvertently) the `${var:-default}` syntax to replace a var explicitly set to the empty string with a default value.

**Comment by faho** [comment]
I have been working on this. The main hurdle I have is that I can't decide if

```fish
set --global foo global

function bar
    set --default --local foo def
end
```

should set $foo locally to "global" or not set it at all. I have found uses for both.
